Sources/SemanticIndex/UpdateIndexStoreTaskDescription.swift

```diff
@@ -0,0 +1,311 @@
+      )
+    }
+  }
+
+  private func updateIndexStore(
+    forSwiftFile uri: DocumentURI,
+    buildSettings: FileBuildSettings,
+    toolchain: Toolchain
+  ) async throws {
+    let indexingArguments = adjustSwiftCompilerArgumentsForIndexStoreUpdate(
+      buildSettings.compilerArguments,
+      fileToIndex: uri
+    )
+
+    guard let swiftc = toolchain.swiftc else {
+      logger.error(
+        "Not updating index store for \(uri.forLogging) because toolchain \(toolchain.identifier) does not contain a Swift compiler"
+      )
+      return
+    }
+
+    let process =
+      if let workingDirectory = buildSettings.workingDirectory {
+        Process(
+          arguments: [swiftc.pathString] + indexingArguments,
+          workingDirectory: try AbsolutePath(validating: workingDirectory)
+        )
+      } else {
+        Process(arguments: [swiftc.pathString] + indexingArguments)
+      }
+    try process.launch()
+    let result = try await process.waitUntilExitSendingSigIntOnTaskCancellation()
+    switch result.exitStatus.exhaustivelySwitchable {
+    case .terminated(code: 0):
+      break
+    case .terminated(code: let code):
+      // This most likely happens if there are compilation errors in the source file. This is nothing to worry about.
+      let stdout = (try? String(bytes: result.output.get(), encoding: .utf8)) ?? "<no stderr>"
+      let stderr = (try? String(bytes: result.stderrOutput.get(), encoding: .utf8)) ?? "<no stderr>"
+      // Indexing will frequently fail if the source code is in an invalid state. Thus, log the failure at a low level.
+      logger.debug(
+        """
+        Updating index store for Swift file \(uri.forLogging) terminated with non-zero exit code \(code)
+        Stderr:
+        \(stderr)
+        Stdout:
+        \(stdout)
+        """
+      )
+      BuildSettingsLogger.log(level: .debug, settings: buildSettings, for: uri)
+    case .signalled(signal: let signal):
+      if !Task.isCancelled {
+        // The indexing job finished with a signal. Could be because the compiler crashed.
+        // Ignore signal exit codes if this task has been cancelled because the compiler exits with SIGINT if it gets
+        // interrupted.
+        logger.error("Updating index store for Swift file \(uri.forLogging) signaled \(signal)")
+        BuildSettingsLogger.log(level: .error, settings: buildSettings, for: uri)
+      }
+    case .abnormal(exception: let exception):
+      if !Task.isCancelled {
+        logger.error("Updating index store for Swift file \(uri.forLogging) exited abnormally \(exception)")
+        BuildSettingsLogger.log(level: .error, settings: buildSettings, for: uri)
+      }
+    }
+  }
+}
+
+/// Adjust compiler arguments that were created for building to compiler arguments that should be used for indexing.
+///
+/// This removes compiler arguments that produce output files and adds arguments to index the file.
+private func adjustSwiftCompilerArgumentsForIndexStoreUpdate(
```

Comment:
This all makes me sad, we really need a "please don't output anything mode". Though... how that would then work with "oh but actually do output the index" is certainly a reasonable question 😅